## Applied Data Science Capstone

### Introduction/Business Problem

The Seattle government is concerned with the severity and number of accidents and wants to employ actions based on the analysis of historical data to alert drivers of the imminence of accidents.

This study aims to predict the severity of an accident, given the locations, weather and road conservation.

The analysis aims to determine a set of possible causes that contribute to the increase in the severity of accidents to allow preventive actions by road users.

The targets audience of the project are drivers, rescue groups, police and insurance companys.

It's expected to achieve a reduction in the number and severity of accidents to make drivers and passengers more secure.

### About dataset

This dataset is about collisions that occurred between 2004 and 2020 in the city of Seattle. The __Data-Collisions.csv__ data set includes details of 194673 collisions provided by the Seattle Department of Transportation Traffic Management Division.

It includes following fields:

| Field          | Description                                                                           |
|----------------|---------------------------------------------------------------------------------------|
| OBJECTID       | Unique identifier                                                                     |
| ADDRTYPE       | Collision address type                                                                |
|                |    (Alley/Block/Intersection)                                                         |
| SEVERITYCODE   | A code that corresponds to the severity of the collision                              |
|                |    (3—fatality / 2b—serious injury / 2—injury / 1—prop damage / 0—unknown)            |         
| (X, Y)         | Spatial characteristics                                                               |
| INTKEY         | Key that corresponds to the intersection associated with a collision                  |
| LOCATION       | Description of the general location of the collision                                  |
| SEVERITYDESC   | A detailed description of the severity of the collision                               |
| COLLISIONTYPE  | Collision type                                                                        |
| PERSONCOUNT    | The total number of people involved in the collision                                  |
| PEDCOUNT       | The number of pedestrians involved in the collision                                   |
| PEDCYLCOUNT    | The number of bicycles involved in the collision                                      |
| VEHCOUNT       | The number of vehicles involved in the collision                                      |
| INCDTTM        | The date and time of the incident                                                     |
| JUNCTIONTYPE	 | Category of junction at which collision took place                                    |
| SDOT_COLCODE	 | A code given to the collision by SDOT                                                 |
| SDOT_COLDESC	 | A description of the collision corresponding to the collision code                    |
| INATTENTIONIND | Whether or not collision was due to inattention                                       |
| UNDERINFL	     | Whether or not a driver involved was under the influence of drugs or alcohol          |
| WEATHER	     | A description of the weather conditions during the time of the collision.             |
| ROADCOND       | The condition of the road during the collision                                        |
| LIGHTCOND	     | The light conditions during the collision                                             |
| PEDROWNOTGRNT  | Whether or not the pedestrian right of way was not granted (Y/N)                      |
| SPEEDING       |	Whether or not speeding was a factor in the collision (Y/N)                          |
| SEGLANEKEY     | A key for the lane segment in which the collision occurred                            |	
| CROSSWALKKEY   | A key for the crosswalk at which the collision occurred                               |
| HITPARKEDCAR   | Whether or not the collision involved hitting a parked car (Y/N)                      |

### Methodology

### Reading the Data 

Downloading the data set and loading the data from the CSV file

In [1]:
import pandas as pd
url = 'https://data-seattlecitygis.opendata.arcgis.com/datasets/5b5c745e0f1f48e7a53acec63a0022ab_0.csv'
df = pd.read_csv(url, low_memory=False)

In [2]:
df.head()

X          Y  OBJECTID  INCKEY  COLDETKEY REPORTNO     STATUS  \
0 -122.344896  47.717173         1    1003       1003  3503158    Matched   
1 -122.376467  47.543774         2   56200      56200  1795087    Matched   
2 -122.360735  47.701487         3  327037     328537  E979380    Matched   
3 -122.297415  47.599233         4  327278     328778  E996362  Unmatched   
4 -122.368001  47.653585         5    1248       1248  3645424  Unmatched   

       ADDRTYPE   INTKEY                                          LOCATION  \
0         Block      NaN    AURORA AVE N BETWEEN N 117TH PL AND N 125TH ST   
1         Block      NaN  35TH AVE SW BETWEEN SW MORGAN ST AND SW HOLLY ST   
2  Intersection  37122.0                        3RD AVE NW AND NW 100TH ST   
3  Intersection  30602.0                M L KING JR WAY S AND S JACKSON ST   
4         Block      NaN       W EWING ST BETWEEN 6TH AVE W AND W EWING PL   

   ... ROADCOND                LIGHTCOND PEDROWNOTGRNT SDOTCOLNUM SPEEDING  \
0  ...      Dry                 Daylight             Y        NaN      NaN   
1  ...      Dry  Dark - Street Lights On           NaN  6015003.0      NaN   
2  ...      Wet                 Daylight           NaN        NaN      NaN   
3  ...      NaN                      NaN           NaN        NaN      NaN   
4  ...      NaN                      NaN           NaN        NaN      NaN   

   ST_COLCODE                              ST_COLDESC  SEGLANEKEY  \
0          45                  Vehicle - Pedalcyclist       15057   
1           0  Vehicle going straight hits pedestrian           0   
2          10                       Entering at angle           0   
3                                                 NaN           0   
4                                                 NaN           0   

   CROSSWALKKEY  HITPARKEDCAR  
0             0             N  
1             0             N  
2             0             N  
3             0             N  
4             0             N  

[5 rows x 40 columns]

In [3]:
df.shape

(221525, 40)

In [4]:
df.dtypes

X                  float64
Y                  float64
OBJECTID             int64
INCKEY               int64
COLDETKEY            int64
REPORTNO            object
STATUS              object
ADDRTYPE            object
INTKEY             float64
LOCATION            object
EXCEPTRSNCODE       object
EXCEPTRSNDESC       object
SEVERITYCODE        object
SEVERITYDESC        object
COLLISIONTYPE       object
PERSONCOUNT          int64
PEDCOUNT             int64
PEDCYLCOUNT          int64
VEHCOUNT             int64
INJURIES             int64
SERIOUSINJURIES      int64
FATALITIES           int64
INCDATE             object
INCDTTM             object
JUNCTIONTYPE        object
SDOT_COLCODE       float64
SDOT_COLDESC        object
INATTENTIONIND      object
UNDERINFL           object
WEATHER             object
ROADCOND            object
LIGHTCOND           object
PEDROWNOTGRNT       object
SDOTCOLNUM         float64
SPEEDING            object
ST_COLCODE          object
ST_COLDESC          object
S

In [6]:
df.describe(include="all")

X              Y       OBJECTID         INCKEY  \
count   214050.000000  214050.000000  221525.000000  221525.000000   
unique            NaN            NaN            NaN            NaN   
top               NaN            NaN            NaN            NaN   
freq              NaN            NaN            NaN            NaN   
mean      -122.330756      47.620196  110763.000000  144824.873233   
std          0.030056       0.056048   63948.903529   89222.567072   
min       -122.419091      47.495573       1.000000    1001.000000   
25%       -122.349280      47.577133   55382.000000   71668.000000   
50%       -122.330363      47.616033  110763.000000  127252.000000   
75%       -122.311998      47.664290  166144.000000  209935.000000   
max       -122.238949      47.734142  221525.000000  334020.000000   

            COLDETKEY REPORTNO   STATUS ADDRTYPE         INTKEY  \
count   221525.000000   221525   221525   217813   71936.000000   
unique            NaN   221522        2        3            NaN   
top               NaN  1782439  Matched    Block            NaN   
freq              NaN        2   195285   144999            NaN   
mean    145053.886627      NaN      NaN      NaN   37626.275537   
std      89598.173976      NaN      NaN      NaN   51968.207339   
min       1001.000000      NaN      NaN      NaN   23807.000000   
25%      71668.000000      NaN      NaN      NaN   28653.000000   
50%     127252.000000      NaN      NaN      NaN   29973.000000   
75%     210155.000000      NaN      NaN      NaN   33984.000000   
max     335520.000000      NaN      NaN      NaN  764413.000000   

                                                 LOCATION  ... ROADCOND  \
count                                              216935  ...   195103   
unique                                              25202  ...        9   
top     BATTERY ST TUNNEL NB BETWEEN ALASKAN WY VI NB ...  ...      Dry   
freq                                                  298  ...   128588   
mean                                                  NaN  ...      NaN   
std                                                   NaN  ...      NaN   
min                                                   NaN  ...      NaN   
25%                                                   NaN  ...      NaN   
50%                                                   NaN  ...      NaN   
75%                                                   NaN  ...      NaN   
max                                                   NaN  ...      NaN   

       LIGHTCOND PEDROWNOTGRNT    SDOTCOLNUM SPEEDING  ST_COLCODE  \
count     194933          5195  1.272050e+05     9929      212112   
unique         9             1           NaN        1          63   
top     Daylight             Y           NaN        Y          32   
freq      119492          5195           NaN     9929       44931   
mean         NaN           NaN  7.971063e+06      NaN         NaN   
std          NaN           NaN  2.611523e+06      NaN         NaN   
min          NaN           NaN  1.007024e+06      NaN         NaN   
25%          NaN           NaN  6.007029e+06      NaN         NaN   
50%          NaN           NaN  8.033011e+06      NaN         NaN   
75%          NaN           NaN  1.018101e+07      NaN         NaN   
max          NaN           NaN  1.307202e+07      NaN         NaN   

                    ST_COLDESC     SEGLANEKEY  CROSSWALKKEY  HITPARKEDCAR  
count                   195212  221525.000000  2.215250e+05        221525  
unique                      62            NaN           NaN             2  
top     One parked--one moving            NaN           NaN             N  
freq                     44931            NaN           NaN        209492  
mean                       NaN     261.135903  9.577244e+03           NaN  
std                        NaN    3246.962907  7.146156e+04           NaN  
min                        NaN       0.000000  0.000000e+00           NaN  
25%                        NaN      

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 221525 entries, 0 to 221524
Data columns (total 40 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   X                214050 non-null  float64
 1   Y                214050 non-null  float64
 2   OBJECTID         221525 non-null  int64  
 3   INCKEY           221525 non-null  int64  
 4   COLDETKEY        221525 non-null  int64  
 5   REPORTNO         221525 non-null  object 
 6   STATUS           221525 non-null  object 
 7   ADDRTYPE         217813 non-null  object 
 8   INTKEY           71936 non-null   float64
 9   LOCATION         216935 non-null  object 
 10  EXCEPTRSNCODE    101122 non-null  object 
 11  EXCEPTRSNDESC    11779 non-null   object 
 12  SEVERITYCODE     221524 non-null  object 
 13  SEVERITYDESC     221525 non-null  object 
 14  COLLISIONTYPE    195212 non-null  object 
 15  PERSONCOUNT      221525 non-null  int64  
 16  PEDCOUNT         221525 non-null  int6

Convert the 'INCDTTM' column to datetime format

In [8]:
df['INCDTTM'] = pd.to_datetime(df['INCDTTM'])
df['INCDATE'] = pd.to_datetime(df['INCDATE'])

In [4]:
df['SEVERITYCODE'].value_counts()

1     137671
2      58783
0      21615
2b      3105
3        350
Name: SEVERITYCODE, dtype: int64

Severity of the collision
3  — fatality: 350
2b — serious injury: 3105
2  — injury: 58783
1  — prop damage: 137671
0  — unknown: 21615

In [9]:
df['ADDRTYPE'].value_counts()

Block           144999
Intersection     71936
Alley              878
Name: ADDRTYPE, dtype: int64

Collision address type
Block:          144999
Intersection:    71936
Alley:             878

In [11]:
df['COLLISIONTYPE'].value_counts()

Parked Car    48551
Angles        35573
Rear Ended    34691
Other         24588
Sideswipe     18891
Left Turn     14115
Pedestrian     7666
Cycles         5932
Right Turn     3017
Head On        2188
Name: COLLISIONTYPE, dtype: int64

Collision type
Parked Car:    48551
Angles:        35573
Rear Ended:    34691
Other:         24588
Sideswipe:     18891
Left Turn:     14115
Pedestrian:     7666
Cycles:         5932
Right Turn:     3017
Head On:        2188

In [12]:
df['INATTENTIONIND'].value_counts()

Y    30188
Name: INATTENTIONIND, dtype: int64

In [13]:
df['UNDERINFL'].value_counts()

N    103927
0     81676
Y      5399
1      4230
Name: UNDERINFL, dtype: int64

In [14]:
df['WEATHER'].value_counts()

Clear                       114738
Raining                      34036
Overcast                     28552
Unknown                      15131
Snowing                        919
Other                          860
Fog/Smog/Smoke                 577
Sleet/Hail/Freezing Rain       116
Blowing Sand/Dirt               56
Severe Crosswind                26
Partly Cloudy                   10
Blowing Snow                     1
Name: WEATHER, dtype: int64

In [15]:
df['ROADCOND'].value_counts()

Dry               128588
Wet                48734
Unknown            15139
Ice                 1232
Snow/Slush          1014
Other                136
Standing Water       119
Sand/Mud/Dirt         77
Oil                   64
Name: ROADCOND, dtype: int64

In [16]:
df['LIGHTCOND'].value_counts()

Daylight                    119492
Dark - Street Lights On      50133
Unknown                      13532
Dusk                          6082
Dawn                          2609
Dark - No Street Lights       1579
Dark - Street Lights Off      1239
Other                          244
Dark - Unknown Lighting         23
Name: LIGHTCOND, dtype: int64

In [17]:
df['SPEEDING'].value_counts()

Y    9929
Name: SPEEDING, dtype: int64

In [18]:
df['SEGLANEKEY'].value_counts()

0        218489
6532         19
6078         19
12162        18
10336        15
          ...  
9803          1
14281         1
4178          1
6355          1
9402          1
Name: SEGLANEKEY, Length: 2101, dtype: int64

In [19]:
df['CROSSWALKKEY'].value_counts()

0         217283
523609        19
520838        15
524265        13
525567        13
           ...  
525111         1
523080         1
521033         1
523208         1
521927         1
Name: CROSSWALKKEY, Length: 2343, dtype: int64